### ISSUE TITLE
z-image batch size ineffective and black samples

### ISSUE THREAD

**Issue by DaveZeRave** [kohya-ss/musubi-tuner#780]
When batch size is changed, training acknowledges "batch size per device / バッチサイズ: 4, 4", but training continues as if there was a batch size of 1, with no change to VRAM usage. Also I'm getting black sample outputs. I think someone has mentioned the --resume epoch issue already. I'm in WSL Python3.10, torch==2.7.0, torchvision==0.22.0, torchaudio==2.7.0, cu128, flash_attn-2.7.4, sageattention-2.1.1.

**Comment by kallu87** [comment]
Yeah same problem here, i am getting black outputs

**Comment by kohya-ss** [comment]
Even if you increase the batch size, if the images have different resolutions, a batch may contain fewer images than the batch size due to Aspect Ratio Bucketing.

I don't think the black images are caused by the batch size. Could you please check if you are using SageAttention for training (it cannot be used for training) and the learning rate etc.? 

It would be good if you could share the entire training command.

**Comment by DaveZeRave** [comment]
Thanks for getting back and all your hard work. Yeah, I know they are most likely two seperate issues, I'm also having the loras not having any effect when tested on a fixed seed issue and epoch/step counts not being passed to --resume.. but they have already been mentioned in other issue threads. I know zit has just been added so to be expexted. I was reading there has been a lot of issues due to the differences between qwen, vs standard text encoders.

I have two sets of codes, tried at both BF16 and FP16 as I wasnt sure if came_pytorch was supported. I also tried completely stripped down versions removing all unnecessary args just to see if there was any conflicts but the issues persisted. "z-image-vae.safetensors" is from Tongyi-MAI, just renamed:

[came_pytorch ver]

accelerate launch --num_cpu_threads_per_process 1 /home/asus/musubi-tuner/src/musubi_tuner/zimage_train_network.py --dit /home/asus/musubi-tuner/models/z_image_de_turbo_v1_bf16.safetensors --vae /home/asus/musubi-tuner/models/z-image-vae.safetensors --text_encoder /home/asus/musubi-tuner/models/qwen_3_4b.safetensors --dataset_config /home/asus/musubi-tuner/toml/zit.toml --flash_attn --mixed_precision fp16 --fp8_base --fp8_scaled --optimizer_type came_pytorch.CAME --optimizer_args weight_decay=0.01 eps=1e-30,1e-16 betas=0.9,0.999,0.9999 --learning_rate 2e-4 --gradient_checkpointing --gradient_accumulation_steps 1 --max_data_loader_n_workers 8 --persistent_data_loader_workers --network_module networks.lora_zimage --network_dim 64 --network_alpha 32 --timestep_sampling shift --weighting_scheme none --discrete_flow_shift 2.0 --max_train_epochs 12000 --save_every_n_epochs 600 --save_state --seed 42 --output_dir /home/asus/musubi-tuner/output/zit-1 --output_name zit-1 --sample_prompts /home/asus/musubi-tuner/prompts/zit-1/prompts.txt --sample_at_first --sample_every_n_epochs 600

[adamw8bit ver]

accelerate launch --num_cpu_threads_per_process 1 /home/asus/musubi-tuner/src/musubi_tuner/zimage_train_network.py --dit /home/asus/musubi-tuner/models/z_image_de_turbo_v1_bf16.safetensors --vae /home/asus/musubi-tuner/models/z-image-vae.safetensors --text_encoder /home/asus/musubi-tuner/models/qwen_3_4b.safetensors --dataset_config /home/asus/musubi-tuner/toml/zit.toml --flash_attn --mixed_precision fp16 --fp8_base --fp8_scaled --optimizer_type adamw8bit --learning_rate 2e-4 --gradient_checkpointing --gradient_accumulation_steps 1 --max_data_loader_n_workers 8 --persistent_data_loader_workers --network_module networks.lora_zimage --network_dim 64 --network_alpha 32 --timestep_sampling shift --weighting_scheme none --discrete_flow_shift 2.0 --max_train_epochs 12000 --save_every_n_epochs 600 --save_state --seed 42 --output_dir /home/asus/musubi-tuner/output/zit-1 --output_name zit-1 --sample_prompts /home/asus/musubi-tuner/prompts/zit-1/prompts.txt --sample_at_first --sample_every_n_epochs 600

WSL 22.04, Python3.10, torch==2.7.0, torchvision==0.22.0, torchaudio==2.7.0, cu128, flash_attn-2.7.4, sageattention-2.1.1. 3090, 128GB DDR4

**Comment by clover-supply** [comment]
Isn't 12 000 steps insane? Aren't the loras just overbaked? Is the output still dead even on early checkpoints around 500-2500 steps?

**Comment by DaveZeRave** [comment]
> Isn't 12 000 steps insane? Aren't the loras just overbaked? Is the output still dead even on early checkpoints around 500-2500 steps?

From my tests so far you have to cook it to get results. I'm not sure if this is to do with fundamental differences to training we are used to, the new encoder/prompting system, that maybe needs more attention to get right.. or just because we're basically training a hacked distilled model.. but as an example. I trained one dataset of 160 images for 30,000 steps at 2e-4 to get it to where I wanted it. 

Early days though, I'm now running more tests and researching into the encoder/prompting side. It could be we are all trying to train using our old prompting methods, and it's just not how it was trained or should be trained... but back to your point, 30,000 steps didn't fry the model. At 12,000 it wasn't even trained..

**Comment by CodeAlexx** [comment]
i been with lokr with ai-toolkit since it came out, convergence happens at around 300 steps for rank 4-6. lr 1e-4, adafactor. Just tried OneTrainer for lora and it does very good. However i came out good my first run, now i use same settings and its not even like my character. I am doing musubi today , at 500 steps no change,-- i had changes at around 500 with onetrainer, normal lora's with ai-toolkit i could not get any likeness. Just frustrating OneTrainer and Musubi does not really change untill make k steps. I did a style lora with Onetrainer and learned the style by 700 steps. Can anyone trainer make consistent loras? Only lokr was the most consistent. 
Just frustration city

**Comment by kohya-ss** [comment]
Since alpha has been changed from the default of 1 to 32, the effective learning rate has been multiplied by 32, which may be causing the training to become unstable. This may be the cause of the black image. Check that the loss is not too large (greater than about 0.5) or NaN. If so, try lowering α or lowering the learning rate to around 1e-5.

Also, is the first sample before training generated correctly and not a black image? If not, the cause may be somewhere other than training.

**Comment by GraftingRayman** [comment]
> Isn't 12 000 steps insane? Aren't the loras just overbaked? Is the output still dead even on early checkpoints around 500-2500 steps?

I have to agree with DaveZeRave on this one, the more images and steps I have, the better the lora seems to be.

I am doing 60-100 images at 10 repeats with 16 epochs and it seems to hit the sweet spot

**Comment by DaveZeRave** [comment]
> > Isn't 12 000 steps insane? Aren't the loras just overbaked? Is the output still dead even on early checkpoints around 500-2500 steps?
> 
> I have to agree with DaveZeRave on this one, the more images and steps I have, the better the lora seems to be.
> 
> I am doing 60-100 images at 10 repeats with 16 epochs and it seems to hit the sweet spot

At 2e-4, I'm not even getting the first stage of convergence until about 8k-9k steps.. I've spent the last couple of days prompting qwen3-4b to try and understand the language it uses natively. My guessing is that training length is most likely to do with using an LLM encoder as they are larger, use causal or bidirectional attention and have longer context windows.. so this multiplies the training time needed. The worst thing you can do to counter this is raise LR like you might with a CLIP.

Things that might work from looking into the differences. Prompt coherence, LLM's punnish much more if there is not a correlative format to training captions. Warm ups, limiting the damage done early in training can massively reduce total training time with LLMs compared to CLIPs. Also CFG dropout can have much stronger effects with an LLM so you should keep between 0-5%. 

30k-40k steps to hit convergence on a 100 image dataset with an LLM-encoder is legit according to A.I.. but for instance, a warmup of 5–10% could drop 10k and give better results.. so is worth it.

**Comment by DaveZeRave** [comment]
> Since alpha has been changed from the default of 1 to 32, the effective learning rate has been multiplied by 32, which may be causing the training to become unstable. This may be the cause of the black image. Check that the loss is not too large (greater than about 0.5) or NaN. If so, try lowering α or lowering the learning rate to around 1e-5.
> 
> Also, is the first sample before training generated correctly and not a black image? If not, the cause may be somewhere other than training.

Thanks for getting back! I thought --network_dim & --network_alpha are usally matched.. Is this something unique about z-image?

**Comment by cvtower** [comment]
With 1epoch lora training with ~3k image-text pairs, i got pretty nice result now.
however, i found z-image might be sensitive to the training prompts.

I will try warmup and other tricks beyond the z-image reference docs, since i dont think a 6B t2i model need 30k~40k steps to converge as mentioned above.

**Comment by DaveZeRave** [comment]
> With 1epoch lora training with ~3k image-text pairs, i got pretty nice result now. however, i found z-image might be sensitive to the training prompts.
> 
> I will try warmup and other tricks beyond the z-image reference docs, since i dont think a 6B t2i model need 30k~40k steps to converge as mentioned above.

The size of the model is irrelevant, as the issue is visual grounding efficiency acting as a bottleneck for LLMs... Optimising for LLM encoding is essential for faster training. 
Here's some estimates on LLM vs CLIP training, and the most important areas to focus on. 

Best-case LLM setup
----------------------------------
Frozen LLM encoder
Short, consistent captions
Low CFG dropout
Warmup
Adequate LoRA rank
➡ ~2×–3× longer than CLIP

Typical real-world setup
----------------------------------
Frozen or lightly-trained LLM
Mixed captions
Default CFG dropout
No warmup
➡ ~4×–6× longer

Worst-case 
----------------------------------
Trainable LLM encoder
High CFG dropout
Narrative captions
Small LoRA rank
No warmup
➡ 8×–10× longer or worse

**Comment by Krishicall** [comment]
Anyone have success with full finetuning? My first attempts also yield models that produce only blackness. I'm using the comfy turbo model with mixed precision bf16.

accelerate launch --num_cpu_threads_per_process 1 src\musubi_tuner\zimage_train.py 
--dit C:\Models\diffusion_models\Z-Image-Turbo-BF16.safetensors 
--text_encoder C:\Models\text_encoders\qwen_3_4b.safetensors 
--vae C:\Models\vae\ae.safetensors
--dataset_config C:\Musubi\musubi-tuner\TOML\ZYnaWindows.toml 
--sdpa --mixed_precision bf16 
--optimizer_type adafactor 
--learning_rate 1e-6 
--fused_backward_pass 
--gradient_checkpointing 
--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False" 
--max_grad_norm 0 
--lr_scheduler constant_with_warmup
--lr_warmup_steps 10 --max_data_loader_n_workers 2 --persistent_data_loader_workers 
--max_train_epochs 300 
--save_every_n_epochs 10 
--save_state 
--mem_eff_save 
--seed 1342569 
--blocks_to_swap 16 
--disable_numpy_memmap 
--use_pinned_memory_for_block_swap 
--logging_dir=logs
